app/frontend/wallet/helpers/CachedDeriveXpubFactory.ts

```diff
@@ -28,14 +37,55 @@ function CachedDeriveXpubFactory(derivationScheme, deriveXpubHardenedFn) {
     return await derivedXpubs[memoKey]
```

Comment:
if we rewrite this to
```
let result
    try {
      result = await derivedXpubs[memoKey]
    } catch (e) {
      delete derivedXpubs[memoKey]
      throw e
    }
```

wouldn't that remove the need for the `cleanXpubCache` function? If I understand correctly, the only purpose of that function is cleaning the failed promises.

And if we stored the actual xpubs instead of the promises, we should be able to get rid of the need to invalidate the cache for failed promises altogether. I'm not sure why I made this solution "resilient" against concurrent execution in the past but if there aren't any apparent issues with removing this "feature" I would not be afraid to do so - worst case it would result in a repeated prompt on the hw wallet, even in the hypothetic case that deriveXpub is called asynchronously multiple times

Reply:
I get it, I thought about this too but I wanted to keep the async behavior. I could have certainly made this easier for me if I dropped the whole logic of storing promises and used the solution you described. Currently the hardened key derivation happens synchronously (which it must because ledger throws error otherwise) and all the other key derivation asynchronously. 

Im not necessarily against dropping the async behavior but I wanted to find a way in which we don't have to. In that case, afaik there is no other way other than cleaning the xpubCache when re-discovering the accounts.